File: rb/test/vocabulary_entity_test.rb
# Vocabulary entity test

require "minitest/autorun"
require "json"
require_relative "../Gbif_sdk"
require_relative "runner"

class VocabularyEntityTest < Minitest::Test
  def test_create_instance
    testsdk = GbifSDK.test(nil, nil)
    ent = testsdk.Vocabulary(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = vocabulary_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "vocabulary." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set GBIF_TEST_VOCABULARY_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    vocabulary_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.vocabulary")))
    vocabulary_ref01_data = nil
    if vocabulary_ref01_data_raw.length > 0
      vocabulary_ref01_data = Helpers.to_map(vocabulary_ref01_data_raw[0][1])
    end

    # LIST
    vocabulary_ref01_ent = client.Vocabulary(nil)
    vocabulary_ref01_match = {}

    vocabulary_ref01_list_result = vocabulary_ref01_ent.list(vocabulary_ref01_match, nil)
    assert vocabulary_ref01_list_result.is_a?(Array)

  end
end

def vocabulary_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "vocabulary", "VocabularyTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = GbifSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["vocabulary01", "vocabulary02", "vocabulary03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["GBIF_TEST_VOCABULARY_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "GBIF_TEST_VOCABULARY_ENTID" => idmap,
    "GBIF_TEST_LIVE" => "FALSE",
    "GBIF_TEST_EXPLAIN" => "FALSE",
    "GBIF_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["GBIF_TEST_VOCABULARY_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["GBIF_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["GBIF_APIKEY"],
      },
      extra || {},
    ])
    client = GbifSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["GBIF_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["GBIF_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/vocabulary_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/gbif-sdk/go"
	"github.com/voxgig-sdk/gbif-sdk/go/core"

	vs "github.com/voxgig-sdk/gbif-sdk/go/utility/struct"
)

func TestVocabularyEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Vocabulary(nil)
		if ent == nil {
			t.Fatal("expected non-nil VocabularyEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"vocabulary": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Vocabulary(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Vocabulary(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := vocabularyBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "vocabulary." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GBIF_TEST_VOCABULARY_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		vocabularyRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.vocabulary", setup.data)))
		var vocabularyRef01Data map[string]any
		if len(vocabularyRef01DataRaw) > 0 {
			vocabularyRef01Data = core.ToMapAny(vocabularyRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = vocabularyRef01Data

		// LIST
		vocabularyRef01Ent := client.Vocabulary(nil)
		vocabularyRef01Match := map[string]any{}

		vocabularyRef01ListResult, err := vocabularyRef01Ent.List(vocabularyRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, vocabularyRef01ListOk := vocabularyRef01ListResult.([]any)
		if !vocabularyRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", vocabularyRef01ListResult)
		}

	})
}

func vocabularyBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "vocabulary", "VocabularyTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read vocabulary test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse vocabulary test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"vocabulary01", "vocabulary02", "vocabulary03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GBIF_TEST_VOCABULARY_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GBIF_TEST_VOCABULARY_ENTID": idmap,
		"GBIF_TEST_LIVE":      "FALSE",
		"GBIF_TEST_EXPLAIN":   "FALSE",
		"GBIF_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["GBIF_TEST_VOCABULARY_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GBIF_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["GBIF_APIKEY"],
			},
			extra,
		})
		client = sdk.NewGbifSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GBIF_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GBIF_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
